sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -706,3 +705,13 @@ case class ShowPartitions(
   override val output: Seq[Attribute] = Seq(
     AttributeReference("partition", StringType, nullable = false)())
 }
+
+/**
+ * The logical plan of the CACHE TABLE command.
+ */
+case class CacheTable(
+    table: LogicalPlan,
+    multipartIdentifier: Seq[String],
+    query: Option[LogicalPlan],
```

Comment:
I also tried to resolve `query` in Analyzer by making it a `child` (similar to `InsertIntoStatement`). However, it has the same issue for v2 tables where `DataSourceV2Relation` is changed to `DataSourceV2ScanRelation` in `V2ScanRelationPushDown`.

Reply:
hmm, for the `query`, it should apply filter pushdown and contain  `DataSourceV2ScanRelation`, right?

Reply:
One issue is invalidating the cache:
```
sql("CACHE TABLE tmp AS SELECT id FROM testcat.ns.tbl")
sql("DROP TABLE testcat.ns.tbl")
```
, where invalidating callback uses `DataSourceV2Relation`: https://github.com/apache/spark/blob/acc211d2cf0e6ab94f6578e1eb488766fd20fa4e/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala#L59-L62

Reply:
> hmm, for the `query`, it should apply filter pushdown and contain `DataSourceV2ScanRelation`, right?

We create a temp view and cache it as follow:
```scala
Dataset.ofRows(sparkSession, query.get).createTempView(relationName)
val df = sparkSession.table(relationName)
sparkSession.sharedState.cacheManager.cacheQuery(df, Some(relationName))
```
`cacheQuery` uses the "analyzed" plan of `df` to cache. The analyzed plan will have `DataSourceV2Relation` (but the optimized plan will have `DataSourceV2ScanRelation`).

If we make the `query` as a child of the plan, `query` will be have an optimized plan `DataSourceV2ScanRelation`.

Reply:
ah i see, thanks for the explaination!